File: 717-1-bit-and-2-bit-characters/717-1-bit-and-2-bit-characters.py
class Solution:
    def isOneBitCharacter(self, bits: List[int]) -> bool:
        #can the last character be two bits
            #two options: 10 and 11
            #can take away 11 off the bat
        '''we must look at bits[:-2] to see if we can fit
           ANY sequence of 0, 10, 11
           recursion: 2^n almost
                try recursive back tracking
           side: if there is a one, it must be two bit
           greedy?
        '''
        if len(bits)==1:
            return True
        if len(bits)==2:
            if bits in [[0,0], [0,1]]:
                return True
            else:
                return False
        if bits[-2:] in [[0,0], [0,1]]: #it cannot end with two bits in these cases
            return True
        
        n = len(bits)-2
        def recursion(i):
            #return true if you reach the end and 
            #we start at i = 0, end at i = len(bits)-2
            '''case 1: i'''
            
            if i == n:
                return False 
            if i >n:
                return True
            if bits[i] == 0:
                return recursion(i+1)
            if bits[i:i+2] in [[1,0], [1,1]]:
                return recursion(i+2)
                    
            else:
                return True
                  
        return recursion(0)        

    
    
    
    
    
    

File: 717-1-bit-and-2-bit-characters/717-1-bit-and-2-bit-characters.java
class Solution {
    public boolean isOneBitCharacter(int[] bits) {
        //determine whether the last character can be a two bit character
        //if so then false
        //so we take two for the last two and try to decode
        /*
            [1,0,0]
            we have so solve recursively
            
            three combos:
                10 or 11 for two bit pairs, or 0 for one bit
                we try decoding as all three and or them
        
        */
        
        if(bits.length == 0)
            return false;
        
        if (bits.length == 1)
            return true;

        if (bits[bits.length-2] == 0 )
            return true;
    
        if (bits.length == 2 &&  bits[0]==1)
            return false;

        if (bits[bits.length-2] == 1 )
            return !recursion(bits, bits.length-3);
        return true;
         
    }
    
    public boolean recursion(int[] bits, int depth){
        if (depth < 0){
            return true;
        }
        if (depth == 0)
            return bits[0]==0;
        else{
            boolean two = false;
            boolean one = false;
            if ( bits[depth-1] == 1 )
                two = recursion(bits, depth-2);
            if ( bits[depth] == 0 )
                one = recursion(bits, depth-1);
            return two || one;
        }
        
        
        
    }
}